File: python/schemapin/pinning.py
"""Key pinning storage and management for Trust-On-First-Use (TOFU)."""

import json
import sqlite3
from pathlib import Path
from typing import Dict, Optional, List
from datetime import datetime


class KeyPinning:
    """Manages key pinning storage using SQLite."""

    def __init__(self, db_path: Optional[str] = None):
        """
        Initialize key pinning storage.
        
        Args:
            db_path: Path to SQLite database file. If None, uses default location.
        """
        if db_path is None:
            db_path = str(Path.home() / '.schemapin' / 'pinned_keys.db')
        
        self.db_path = db_path
        self._ensure_db_directory()
        self._init_database()

    def _ensure_db_directory(self) -> None:
        """Ensure database directory exists."""
        Path(self.db_path).parent.mkdir(parents=True, exist_ok=True)

    def _init_database(self) -> None:
        """Initialize database schema."""
        with sqlite3.connect(self.db_path) as conn:
            conn.execute('''
                CREATE TABLE IF NOT EXISTS pinned_keys (
                    tool_id TEXT PRIMARY KEY,
                    public_key_pem TEXT NOT NULL,
                    domain TEXT NOT NULL,
                    developer_name TEXT,
                    pinned_at TEXT NOT NULL,
                    last_verified TEXT
                )
            ''')
            conn.commit()

    def pin_key(
        self, 
        tool_id: str, 
        public_key_pem: str, 
        domain: str,
        developer_name: Optional[str] = None
    ) -> bool:
        """
        Pin a public key for a tool.
        
        Args:
            tool_id: Unique tool identifier
            public_key_pem: PEM-encoded public key
            domain: Tool provider domain
            developer_name: Optional developer name
            
        Returns:
            True if key was pinned successfully, False if already exists
        """
        try:
            with sqlite3.connect(self.db_path) as conn:
                conn.execute('''
                    INSERT INTO pinned_keys 
                    (tool_id, public_key_pem, domain, developer_name, pinned_at)
                    VALUES (?, ?, ?, ?, ?)
                ''', (
                    tool_id, 
                    public_key_pem, 
                    domain, 
                    developer_name,
                    datetime.utcnow().isoformat()
                ))
                conn.commit()
                return True
        except sqlite3.IntegrityError:
            return False

    def get_pinned_key(self, tool_id: str) -> Optional[str]:
        """
        Get pinned public key for tool.
        
        Args:
            tool_id: Tool identifier
            
        Returns:
            PEM-encoded public key if pinned, None otherwise
        """
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.execute(
                'SELECT public_key_pem FROM pinned_keys WHERE tool_id = ?',
                (tool_id,)
            )
            result = cursor.fetchone()
            return result[0] if result else None

    def is_key_pinned(self, tool_id: str) -> bool:
        """
        Check if key is pinned for tool.
        
        Args:
            tool_id: Tool identifier
            
        Returns:
            True if key is pinned, False otherwise
        """
        return self.get_pinned_key(tool_id) is not None

    def update_last_verified(self, tool_id: str) -> bool:
        """
        Update last verification timestamp for tool.
        
        Args:
            tool_id: Tool identifier
            
        Returns:
            True if updated successfully, False if tool not found
        """
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.execute('''
                UPDATE pinned_keys 
                SET last_verified = ? 
                WHERE tool_id = ?
            ''', (datetime.utcnow().isoformat(), tool_id))
            conn.commit()
            return cursor.rowcount > 0

    def list_pinned_keys(self) -> List[Dict[str, str]]:
        """
        List all pinned keys with metadata.
        
        Returns:
            List of dictionaries containing key information
        """
        with sqlite3.connect(self.db_path) as conn:
            conn.row_factory = sqlite3.Row
            cursor = conn.execute('''
                SELECT tool_id, domain, developer_name, pinned_at, last_verified
                FROM pinned_keys
                ORDER BY pinned_at DESC
            ''')
            return [dict(row) for row in cursor.fetchall()]

    def remove_pinned_key(self, tool_id: str) -> bool:
        """
        Remove pinned key for tool.
        
        Args:
            tool_id: Tool identifier
            
        Returns:
            True if removed successfully, False if not found
        """
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.execute(
                'DELETE FROM pinned_keys WHERE tool_id = ?',
                (tool_id,)
            )
            conn.commit()
            return cursor.rowcount > 0

    def get_key_info(self, tool_id: str) -> Optional[Dict[str, str]]:
        """
        Get complete information about pinned key.
        
        Args:
            tool_id: Tool identifier
            
        Returns:
            Dictionary with key information if found, None otherwise
        """
        with sqlite3.connect(self.db_path) as conn:
            conn.row_factory = sqlite3.Row
            cursor = conn.execute('''
                SELECT tool_id, public_key_pem, domain, developer_name, 
                       pinned_at, last_verified
                FROM pinned_keys 
                WHERE tool_id = ?
            ''', (tool_id,))
            result = cursor.fetchone()
            return dict(result) if result else None

    def export_pinned_keys(self) -> str:
        """
        Export all pinned keys to JSON format.
        
        Returns:
            JSON string containing all pinned keys
        """
        keys = self.list_pinned_keys()
        # Add public key data for export
        for key_info in keys:
            key_info['public_key_pem'] = self.get_pinned_key(key_info['tool_id'])
        
        return json.dumps(keys, indent=2)

    def import_pinned_keys(self, json_data: str, overwrite: bool = False) -> int:
        """
        Import pinned keys from JSON format.
        
        Args:
            json_data: JSON string containing key data
            overwrite: Whether to overwrite existing keys
            
        Returns:
            Number of keys imported
        """
        try:
            keys = json.loads(json_data)
            imported = 0
            
            for key_info in keys:
                if overwrite and self.is_key_pinned(key_info['tool_id']):
                    self.remove_pinned_key(key_info['tool_id'])
                
                if self.pin_key(
                    key_info['tool_id'],
                    key_info['public_key_pem'],
                    key_info['domain'],
                    key_info.get('developer_name')
                ):
                    imported += 1
            
            return imported
        except (json.JSONDecodeError, KeyError):
            return 0

File: javascript/src/pinning.js
/**
 * Key pinning storage and management for Trust-On-First-Use (TOFU).
 */

import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'fs';
import { join, dirname } from 'path';
import { homedir } from 'os';

/**
 * Manages key pinning storage using JSON file.
 */
export class KeyPinning {
    /**
     * Initialize key pinning storage.
     * 
     * @param {string|null} dbPath - Path to JSON storage file. If null, uses default location.
     */
    constructor(dbPath = null) {
        if (dbPath === null) {
            const homeDir = homedir();
            dbPath = join(homeDir, '.schemapin', 'pinned_keys.json');
        }
        
        this.dbPath = dbPath;
        this._ensureDbDirectory();
        this._initDatabase();
    }

    /**
     * Ensure database directory exists.
     * @private
     */
    _ensureDbDirectory() {
        const dir = dirname(this.dbPath);
        if (!existsSync(dir)) {
            mkdirSync(dir, { recursive: true });
        }
    }

    /**
     * Initialize database file.
     * @private
     */
    _initDatabase() {
        if (!existsSync(this.dbPath)) {
            this._saveData({});
        }
    }

    /**
     * Load data from JSON file.
     * @private
     * @returns {Object} Pinned keys data
     */
    _loadData() {
        try {
            const data = readFileSync(this.dbPath, 'utf8');
            return JSON.parse(data);
        } catch (error) {
            return {};
        }
    }

    /**
     * Save data to JSON file.
     * @private
     * @param {Object} data - Data to save
     */
    _saveData(data) {
        writeFileSync(this.dbPath, JSON.stringify(data, null, 2), 'utf8');
    }

    /**
     * Pin a public key for a tool.
     * 
     * @param {string} toolId - Unique tool identifier
     * @param {string} publicKeyPem - PEM-encoded public key
     * @param {string} domain - Tool provider domain
     * @param {string|null} developerName - Optional developer name
     * @returns {boolean} True if key was pinned successfully, false if already exists
     */
    pinKey(toolId, publicKeyPem, domain, developerName = null) {
        const data = this._loadData();
        
        if (toolId in data) {
            return false; // Key already exists
        }
        
        data[toolId] = {
            public_key_pem: publicKeyPem,
            domain: domain,
            developer_name: developerName,
            pinned_at: new Date().toISOString(),
            last_verified: null
        };
        
        this._saveData(data);
        return true;
    }

    /**
     * Get pinned public key for tool.
     * 
     * @param {string} toolId - Tool identifier
     * @returns {string|null} PEM-encoded public key if pinned, null otherwise
     */
    getPinnedKey(toolId) {
        const data = this._loadData();
        const keyInfo = data[toolId];
        return keyInfo ? keyInfo.public_key_pem : null;
    }

    /**
     * Check if key is pinned for tool.
     * 
     * @param {string} toolId - Tool identifier
     * @returns {boolean} True if key is pinned, false otherwise
     */
    isKeyPinned(toolId) {
        return this.getPinnedKey(toolId) !== null;
    }

    /**
     * Update last verification timestamp for tool.
     * 
     * @param {string} toolId - Tool identifier
     * @returns {boolean} True if updated successfully, false if tool not found
     */
    updateLastVerified(toolId) {
        const data = this._loadData();
        
        if (!(toolId in data)) {
            return false;
        }
        
        data[toolId].last_verified = new Date().toISOString();
        this._saveData(data);
        return true;
    }

    /**
     * List all pinned keys with metadata.
     * 
     * @returns {Array} Array of objects containing key information
     */
    listPinnedKeys() {
        const data = this._loadData();
        return Object.entries(data)
            .map(([toolId, keyInfo]) => ({
                tool_id: toolId,
                domain: keyInfo.domain,
                developer_name: keyInfo.developer_name,
                pinned_at: keyInfo.pinned_at,
                last_verified: keyInfo.last_verified
            }))
            .sort((a, b) => new Date(b.pinned_at) - new Date(a.pinned_at));
    }

    /**
     * Remove pinned key for tool.
     * 
     * @param {string} toolId - Tool identifier
     * @returns {boolean} True if removed successfully, false if not found
     */
    removePinnedKey(toolId) {
        const data = this._loadData();
        
        if (!(toolId in data)) {
            return false;
        }
        
        delete data[toolId];
        this._saveData(data);
        return true;
    }

    /**
     * Get complete information about pinned key.
     * 
     * @param {string} toolId - Tool identifier
     * @returns {Object|null} Object with key information if found, null otherwise
     */
    getKeyInfo(toolId) {
        const data = this._loadData();
        const keyInfo = data[toolId];
        
        if (!keyInfo) {
            return null;
        }
        
        return {
            tool_id: toolId,
            public_key_pem: keyInfo.public_key_pem,
            domain: keyInfo.domain,
            developer_name: keyInfo.developer_name,
            pinned_at: keyInfo.pinned_at,
            last_verified: keyInfo.last_verified
        };
    }

    /**
     * Export all pinned keys to JSON format.
     * 
     * @returns {string} JSON string containing all pinned keys
     */
    exportPinnedKeys() {
        const keys = this.listPinnedKeys();
        // Add public key data for export
        keys.forEach(keyInfo => {
            keyInfo.public_key_pem = this.getPinnedKey(keyInfo.tool_id);
        });
        
        return JSON.stringify(keys, null, 2);
    }

    /**
     * Import pinned keys from JSON format.
     * 
     * @param {string} jsonData - JSON string containing key data
     * @param {boolean} overwrite - Whether to overwrite existing keys
     * @returns {number} Number of keys imported
     */
    importPinnedKeys(jsonData, overwrite = false) {
        try {
            const keys = JSON.parse(jsonData);
            let imported = 0;
            
            for (const keyInfo of keys) {
                if (overwrite && this.isKeyPinned(keyInfo.tool_id)) {
                    this.removePinnedKey(keyInfo.tool_id);
                }
                
                if (this.pinKey(
                    keyInfo.tool_id,
                    keyInfo.public_key_pem,
                    keyInfo.domain,
                    keyInfo.developer_name
                )) {
                    imported++;
                }
            }
            
            return imported;
        } catch (error) {
            return 0;
        }
    }
}